src/Raven.Server/Documents/Indexes/Index.cs

```diff
@@ -1385,23 +1389,51 @@ internal void HandleExcessiveNumberOfReduceErrors(IndexingStatsScope stats, Exce
             if (State == IndexState.Error)
                 return;
 
-            _errorStateReason = e.Message;
-            SetState(IndexState.Error, ignoreWriteError: true);
+            SetErrorState(e.Message);
         }
 
-        internal void HandleDiskFullErrors(IndexingStatsScope stats, DiskFullException dfe)
+        internal void HandleDiskFullErrors(IndexingStatsScope stats, StorageEnvironment storageEnvironment, DiskFullException dfe)
         {
-            if (_logger.IsOperationsEnabled)
-                _logger.Operations($"Disk full error occurred for '{Name}'.", dfe);
-
             stats.AddDiskFullError(dfe);
 
             var diskFullErrors = Interlocked.Increment(ref _diskFullErrors);
+            if (diskFullErrors < DiskFullErrorLimit)
+            {
+                var timeToWaitInSeconds = (int)Math.Min(Math.Pow(2, diskFullErrors), 30);
+
+                if (_logger.IsOperationsEnabled)
+                    _logger.Operations($"After disk full error in index : '{Name}', " +
+                                       $"going to try flushing and syncing the environment to cleanup the storage. " +
+                                       $"Will wait for flush for: {timeToWaitInSeconds} seconds", dfe);
 
-            if (State == IndexState.Error || diskFullErrors < DiskFullErrorLimit)
+                // force flush and sync
+                GlobalFlushingBehavior.GlobalFlusher.Value.MaybeFlushEnvironment(storageEnvironment);
+                if (_logsAppliedEvent.Wait(timeToWaitInSeconds))
+                {
+                    storageEnvironment.ForceSyncDataFile();
+                }
+
+                // wait for sync
+                Task.Delay(1000 * timeToWaitInSeconds, _indexingProcessCancellationTokenSource.Token).Wait();
```

Comment:
Why are you doing this? We are effectively waiting twice.
In any case, worse case we can try again and we'll wait for the sync on the next iteration, no?

Reply:
fixed the calculations